core/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java

```diff
@@ -79,31 +81,16 @@ void setChunkSize(ByteSizeValue chunkSize) { // only settable for tests
     }
 
     public void resync(IndexShard indexShard, ActionListener<ResyncTask> listener) {
-        final Translog.View view = indexShard.acquireTranslogView();
-        ActionListener<ResyncTask> wrappedListener = new ActionListener<ResyncTask>() {
-            @Override
-            public void onResponse(ResyncTask resyncTask) {
-                try {
-                    view.close();
-                } catch (IOException e) {
-                    onFailure(e);
-                }
-                listener.onResponse(resyncTask);
-            }
-
-            @Override
-            public void onFailure(Exception e) {
-                try {
-                    view.close();
-                } catch (IOException inner) {
-                    e.addSuppressed(inner);
-                }
-                listener.onFailure(e);
-            }
-        };
         try {
             final long startingSeqNo = indexShard.getGlobalCheckpoint() + 1;
-            Translog.Snapshot snapshot = view.snapshot(startingSeqNo);
+            Translog.Snapshot snapshot = indexShard.getTranslog().newSnapshotFromMinSeqNo(startingSeqNo);
+            listener = chain(wrap(r -> snapshot.close(), e -> {
```

Comment:
I disagree it allows focusing on the functionality, to chase down what is really being executed here you have to run off and grok two methods, it's really not straightforward at all. With the listener defined front and center you can immediately see what is happening, no need to chase anything down. Sorry, I feel very strongly about this one.